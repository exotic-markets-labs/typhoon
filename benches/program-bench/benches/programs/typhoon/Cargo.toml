[package]
name = "typhoon"
version = "0.1.0"
edition.workspace = true
publish.workspace = true

[features]
logging = []

[lib]
crate-type = ["cdylib", "lib"]

[lints.rust]
unexpected_cfgs = { level = "warn", check-cfg = [
    'cfg(target_os, values("solana"))',
] }

[dependencies]
bytemuck = { version = "1.23", features = ["derive"] }
typhoon = { path = "../../../../../crates/lib" }
