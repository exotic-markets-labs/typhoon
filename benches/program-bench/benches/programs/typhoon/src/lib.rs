#![no_std]

use {
    bytemuck::{AnyBitPattern, NoUninit},
    typhoon::prelude::*,
};

nostd_panic_handler!();
no_allocator!();

program_id!("Bench111111111111111111111111111111111111111");

handlers! {
    ping,
    log,
    create_account,
    transfer
}

fn ping() -> ProgramResult {
    Ok(())
}

fn log() -> ProgramResult {
    msg!("Instruction: Log");
    Ok(())
}

fn create_account(ctx: CreateAccountContext) -> ProgramResult {
    ctx.account.mut_data()?.byte = 1;

    Ok(())
}

fn transfer(Arg(amount): Arg<[u8; 8]>, ctx: TransferContext) -> ProgramResult {
    ctx.admin
        .transfer(&ctx.account, u64::from_le_bytes(*amount))
}

#[context]
pub struct CreateAccountContext {
    pub admin: Mut<Signer>,
    #[constraint(
        init,
        payer = admin
    )]
    pub account: Mut<Account<Data>>,
    pub system_program: Program<System>,
}

#[context]
pub struct TransferContext {
    pub admin: Mut<Signer>,
    pub account: Mut<SystemAccount>,
    pub system_program: Program<System>,
}

#[derive(NoUninit, AnyBitPattern, AccountState, Copy, Clone)]
#[repr(C)]
pub struct Data {
    pub byte: u8,
}
