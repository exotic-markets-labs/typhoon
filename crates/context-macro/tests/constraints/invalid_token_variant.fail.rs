use typhoon::prelude::*;

program_id!("Fg6PaFpoGXkYsidMpWTK6W2BeZ7FEfcYkg476zPFsLnS");

#[account]
pub struct Counter {
    pub count: u64,
}

#[context]
pub struct InitContext {
    pub mint: Mut<Signer>,
    #[constraint(
        token::owner = mint
    )]
    pub counter: Mut<Account<TokenAccount>>,
}

pub fn main() {}
