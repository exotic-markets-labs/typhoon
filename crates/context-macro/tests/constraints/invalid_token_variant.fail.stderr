error: Invalid variant for the token constraint.
  --> tests/constraints/invalid_token_variant.fail.rs:23:26
   |
23 |         token::authority = mint
   |                          ^
